driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -96,108 +113,137 @@ public void connect() {
+                    currentCheckCancelled = false;
+                    waitForNext();
+                    continue;
+                }
+
+                logStateChange(previousServerDescription, currentServerDescription);
+                serverStateListener.stateChanged(new ChangeEvent<>(previousServerDescription, currentServerDescription));
+
+                if (currentServerDescription.getException() != null) {
+                    connectionPool.invalidate();
+                }
+
+                if (currentServerDescription.getTopologyVersion() != null
+                        || (connection != null && connection.hasMoreToCome())
+                        || (currentServerDescription.getException() instanceof MongoSocketException
+                        && previousServerDescription.getType() != UNKNOWN)) {
+                    continue;
                 }
+                waitForNext();
             }
         }
 
         private ServerDescription getConnectingServerDescription(final Throwable exception) {
             return ServerDescription.builder().type(UNKNOWN).state(CONNECTING).address(serverId.getAddress()).exception(exception).build();
         }
 
-        private ServerDescription lookupServerDescription(final InternalConnection connection) {
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug(format("Checking status of %s", serverId.getAddress()));
-            }
-            serverMonitorListener.serverHearbeatStarted(new ServerHeartbeatStartedEvent(connection.getDescription().getConnectionId()));
-
-            long start = System.nanoTime();
+        private ServerDescription lookupServerDescription(final ServerDescription currentServerDescription) {
             try {
-                BsonDocument isMasterResult =
-                        executeCommand("admin", new BsonDocument("ismaster", new BsonInt32(1)), clusterClock, connection);
-                long elapsedTimeNanos = System.nanoTime() - start;
-                averageRoundTripTime.addSample(elapsedTimeNanos);
-
-                serverMonitorListener.serverHeartbeatSucceeded(
-                        new ServerHeartbeatSucceededEvent(connection.getDescription().getConnectionId(), isMasterResult, elapsedTimeNanos));
-
-                return createServerDescription(serverId.getAddress(), isMasterResult, averageRoundTripTime.getAverage());
-            } catch (RuntimeException e) {
-                serverMonitorListener.serverHeartbeatFailed(
-                        new ServerHeartbeatFailedEvent(connection.getDescription().getConnectionId(), System.nanoTime() - start, e));
-                throw e;
+                if (connection == null || connection.isClosed()) {
+                    connection = internalConnectionFactory.create(serverId);
+                    connection.open();
+                    averageRoundTripTime.addSample(connection.getInitialServerDescription().getRoundTripTimeNanos());
```

Comment:
Does the `averageRoundTripTime` still get reset on isMaster/network errors?

Reply:
It doesn't.  Good catch.  Think we should add this sort of thing to the pseudocode?

Reply:
Yes. Opened: https://jira.mongodb.org/browse/SPEC-1690

Reply:
Done.  Still need to do the test though

Reply:
Is this test now done?

Reply:
Yes, the prose test has been implemented.